linkerd/drain/src/lib.rs

```diff
@@ -1,95 +1,97 @@
-};
-use tokio::{
-    stream::Stream,
-    sync::{mpsc, watch},
-};
+use std::future::Future;
+use tokio::sync::{mpsc, watch};
 
 /// Creates a drain channel.
 ///
 /// The `Signal` is used to start a drain, and the `Watch` will be notified
 /// when a drain is signaled.
 pub fn channel() -> (Signal, Watch) {
-    let (tx, rx) = watch::channel(());
+    let (signal_tx, signal_rx) = watch::channel(());
     let (drained_tx, drained_rx) = mpsc::channel(1);
 
-    (Signal { drained_rx, tx }, Watch { drained_tx, rx })
+    let signal = Signal {
+        drained_rx,
+        signal_tx,
+    };
+    let watch = Watch {
+        drained_tx,
+        signal_rx,
+    };
+    (signal, watch)
 }
 
+enum Never {}
+
 /// Send a drain command to all watchers.
-///
-/// When a drain is started, this returns a `Drained` future which resolves
-/// when all `Watch`ers have been dropped.
 #[derive(Debug)]
 pub struct Signal {
     drained_rx: mpsc::Receiver<Never>,
-    tx: watch::Sender<()>,
+    signal_tx: watch::Sender<()>,
 }
 
 /// Watch for a drain command.
 ///
-/// This wraps another future and callback to be called when drain is triggered.
-#[pin_project]
+/// All `Watch` instances must be dropped for a `Signal::signal` call to
+/// complete.
 #[derive(Clone, Debug)]
 pub struct Watch {
-    #[pin]
     drained_tx: mpsc::Sender<Never>,
-    rx: watch::Receiver<()>,
-}
-
-/// A future that resolves when all `Watch`ers have been dropped (drained).
-#[pin_project]
-pub struct Drained {
-    #[pin]
-    drained_rx: mpsc::Receiver<Never>,
+    signal_rx: watch::Receiver<()>,
 }
 
 #[must_use = "ReleaseShutdown should be dropped explicitly to release the runtime"]
 #[derive(Clone, Debug)]
 pub struct ReleaseShutdown(mpsc::Sender<Never>);
 
-// ===== impl Signal =====
+// === impl Signal ===
 
 impl Signal {
     /// Start the draining process.
     ///
     /// A signal is sent to all futures watching for the signal. A new future
     /// is returned from this method that resolves when all watchers have
     /// completed.
-    pub fn drain(self) -> Drained {
-        let _ = self.tx.send(());
-        Drained {
-            drained_rx: self.drained_rx,
+    pub async fn signal(mut self) {
+        // Update the state of the signal watch so that all watchers are observe
+        // the change.
+        let _ = self.signal_tx.send(());
+
+        // Wait for all watchers to release their drain handle.
+        match self.drained_rx.recv().await {
+            None => {}
+            Some(n) => match n {},
         }
     }
 }
 
-// ===== impl Watch =====
+// === impl Watch ===
 
 impl Watch {
     /// Returns a `ReleaseShutdown` handle after the drain has been signaled. The
     /// handle must be dropped when a shutdown action has been completed to
     /// unblock graceful shutdown.
-    pub async fn signal(mut self) -> ReleaseShutdown {
-        let _ = self.rx.changed().await;
+    pub async fn signaled(mut self) -> ReleaseShutdown {
```

Comment:
+1 for this name change even if the `Drain` method isn't renamed to `signal` --- in Tokio, Carl suggested a style where an async fn that represents some event/condition has names ending with `ed`, like `Notify::notified`, `Watch::changed`, etc...